pkg/kvstore/allocator/idpool.go

```diff
@@ -0,0 +1,281 @@
+// pool, or the pool has since been refreshed.
+//
+// Returns true if the ID was returned back to the pool as
+// a result of this call.
+func (p *idPool) Release(id ID) bool {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	return p.idCache.release(id)
+}
+
+// Use makes a leased ID unavailable in the pool and has no effect
+// otherwise. Returns true if the ID was made unavailable
+// as a result of this call.
+func (p *idPool) Use(id ID) bool {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	return p.idCache.use(id)
+}
+
+// Insert makes an unavailable ID available in the pool
+// and has no effect otherwise. Returns true if the ID
+// was added back to the pool.
+func (p *idPool) Insert(id ID) bool {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	return p.nextIDCache.insert(id)
+}
+
+// Remove makes an ID unavailable in the pool.
+// Returns true if the ID was previously available in the pool.
+func (p *idPool) Remove(id ID) bool {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	return p.nextIDCache.remove(id)
+}
+
+type idCache struct {
+
+	// ids is a slice of IDs available in this idCache.
+	ids []ID
+
+	// index tracks the position of IDs in the above ids slice.
+	index map[ID]int
+
+	// leased is a slice of IDs that are leased in this idCache.
+	leased map[ID]bool
+}
+
+func newIDCache(minID ID, maxID ID) *idCache {
+	N := int(maxID - minID + 1)
+	if N < 0 {
+		N = 0
+	}
+
+	c := &idCache{
+		ids:    make([]ID, N),
+		index:  make(map[ID]int),
+		leased: make(map[ID]bool),
+	}
+
+	for i := 0; i < N; i++ {
+		id := ID(i) + minID
+		c.ids[i] = id
+		c.index[id] = i
+	}
+
+	return c
+}
+
+var random = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+// selectAvailableID returns a random available ID.
+func (c *idCache) selectAvailableID() ID {
+	if len(c.ids) == 0 {
+		return NoID
+	}
+
+	id := c.ids[random.Intn(len(c.ids))]
+	c.doRemove(id)
+	// Mark the selected ID as leased.
+	c.leased[id] = true
+
+	return id
+}
+
+// release makes the ID available again if it is currently
+// leased and has no effect otherwise. Returns true if the
+// ID was made available as a result of this call.
+func (c *idCache) release(id ID) bool {
+	if _, exists := c.leased[id]; !exists {
+		return false
+	}
+
+	delete(c.leased, id)
+	c.insert(id)
+
+	return true
+}
+
+// use makes the ID unavailable if it is currently
+// leased and has no effect otherwise. Returns true if the
+// ID was made unavailable as a result of this call.
+func (c *idCache) use(id ID) bool {
+	if _, exists := c.leased[id]; !exists {
+		return false
+	}
+
+	delete(c.leased, id)
+	return true
+}
+
+// insert makes the ID available by adding it back
+// to the cache if it is currently unavailable.
+// Returns true if the ID was added to the cache.
+func (c *idCache) insert(id ID) bool {
+	if _, exists := c.index[id]; exists {
+		return false
+	}
+
+	if _, exists := c.leased[id]; exists {
+		return false
+	}
+
+	i := len(c.ids)
+	c.ids = c.ids[:i+1]
```

Comment:
This doesn't do what you think. If the slice is already at capacity (e.g. `len(c.ids) == N`), this panics.
Use `append` instead:
```
c.ids = append(c.ids, id)
```

And add a unit test to insert an ID into a freshly created `idCache` with all its N ID still available. This shouldn't panic anymore.

Reply:
Done :) The previous intent was to avoid allocating a new array based on the assumption that if an inserted ID falls beyond the min,max range, then the caller's code is faulty or the pool is being misused.